File: 768-partition-labels/partition-labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        Idx = {}
        for i, c in enumerate(s):
            Idx[c] = i
        
        res = []
        st, en = 0, 0

        for i, c in enumerate(s):
            st += 1
            if Idx[c] > en:
                en = Idx[c]
            if i == en:
                res.append(st)
                st = 0
        
        return res

File: 768-partition-labels/partition-labels.java
class Solution {
    public List<Integer> partitionLabels(String s) {
        int[] Idx = new int[26];
        for (int i = 0; i < s.length(); i++) {
            Idx[s.charAt(i) - 'a'] = i; // to convert ch ascii to 0 - 25 range
        }

        List<Integer> res = new ArrayList<>();
        int st = 0;
        int en = 0;

        for (int i = 0; i < s.length(); i++) {
            st++;

            en = Math.max(en, Idx[s.charAt(i) - 'a']);

            if (i == en) {
                res.add(st);
                st = 0;  // Reset st for the next partition.
            }
        }

        return res;
    }
}